File: Editorial/1062_Longest_Repeating_Substring/Python3/3_Suffix_Array_with_Sorting.py
class Solution:
    def longestRepeatingSubstring(self, s: str) -> int:
        length = len(s)
        suffixes = []

        # Create suffix array by storing all suffixes of the string
        for i in range(length):
            suffixes.append(s[i:])
        # Sort the suffix array
        suffixes.sort()

        max_length = 0
        # Compare adjacent suffixes to find the longest common prefix
        for i in range(1, length):
            j = 0
            # Compare characters one by one until
            # they differ or end of one suffix is reached
            while (
                j < min(len(suffixes[i]), len(suffixes[i - 1]))
                and suffixes[i][j] == suffixes[i - 1][j]
            ):
                j += 1
            # Update max_length with the length of the common prefix
            max_length = max(max_length, j)
        return max_length

File: Editorial/1001-1100/1062_Longest_Repeating_Substring/CPP/3_Suffix_Array_with_Sorting.cpp
class Solution {
public:
    int longestRepeatingSubstring(string s) {
        int length = s.length();
        vector<string> suffixes(length);

        // Create suffix array by storing all suffixes of the string
        for (int i = 0; i < length; i++) {
            suffixes[i] = s.substr(i);
        }
        // Sort the suffix array
        sort(suffixes.begin(), suffixes.end());

        int maxLength = 0;
        // Compare adjacent suffixes to find the longest common prefix
        for (int i = 1; i < length; i++) {
            int j = 0;
            // Compare characters one by one until they differ or end of one
            // suffix is reached
            while (j < min(suffixes[i].length(), suffixes[i - 1].length()) &&
                   suffixes[i][j] == suffixes[i - 1][j]) {
                j++;
            }
            // Update maxLength with the length of the common prefix
            maxLength = max(maxLength, j);
        }
        return maxLength;
    }
};
